File: 0287-find-the-duplicate-number/solution.cpp
class Solution {
public:
    int findDuplicate(vector<int>& nums) 
    {   
        
        int tortoise, rabbit;
        tortoise = nums[0];
        rabbit   = nums[0];
        
        do
        {
            tortoise = nums[tortoise];
            rabbit   = nums[nums[rabbit]];
            
        }while(tortoise != rabbit);

        tortoise = nums[0];
        
        while(tortoise != rabbit)
        {
            tortoise = nums[tortoise];
            rabbit   = nums[rabbit];
        }
        
        return tortoise;
    }
};

File: 0287-find-the-duplicate-number/solution.py
class Solution(object):
    def findDuplicate(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        t = nums[0]
        h = nums[0]
        
        while True:
            t = nums[t]
            h = nums[nums[h]]
            if t == h:
                break
            
        t = nums[0]
        while t != h:
            t = nums[t]
            h = nums[h]
        
        return h
